src/coreclr/src/vm/comdelegate.cpp

```diff
@@ -789,41 +789,32 @@ static PCODE GetVirtualCallStub(MethodDesc *method, TypeHandle scopeType)
     return pTargetCall;
 }
 
-FCIMPL5(FC_BOOL_RET, COMDelegate::BindToMethodName,
-                        Object *refThisUNSAFE,
-                        Object *targetUNSAFE,
-                        ReflectClassBaseObject *pMethodTypeUNSAFE,
-                        StringObject* methodNameUNSAFE,
-                        int flags)
+BOOL QCALLTYPE COMDelegate::BindToMethodName(
+                    QCall::ObjectHandleOnStack delegate,
+                    QCall::ObjectHandleOnStack target,
+                    QCall::TypeHandle methodTypeHandle,
+                    PCWSTR pwzMethodName,
+                    int flags)
 {
-    FCALL_CONTRACT;
-
-    struct _gc
-    {
-        DELEGATEREF refThis;
-        OBJECTREF target;
-        STRINGREF methodName;
-        REFLECTCLASSBASEREF refMethodType;
-    } gc;
+    QCALL_CONTRACT;
 
-    gc.refThis    = (DELEGATEREF) ObjectToOBJECTREF(refThisUNSAFE);
-    gc.target     = (OBJECTREF) targetUNSAFE;
-    gc.methodName = (STRINGREF) methodNameUNSAFE;
-    gc.refMethodType = (REFLECTCLASSBASEREF) ObjectToOBJECTREF(pMethodTypeUNSAFE);
+    MethodDesc *pMatchingMethod = NULL;
 
-    TypeHandle methodType = gc.refMethodType->GetType();
+    BEGIN_QCALL;
+    GCX_COOP();
 
-    MethodDesc *pMatchingMethod = NULL;
+    DELEGATEREF delegateRef   = (DELEGATEREF) delegate.Get();
```

Comment:
You still have to protect the GC references like before. Number of the calls below triggers GC - it is not ok to have unprotected GC reference in a local while calling GC triggering method.

The easiest way to do this is to keep the gc struct and use `GCPROTECT_BEGIN/END` macro to protect it.

It may be possible to avoid `GCPROTECT_BEGIN/END` and instead re-fetch the value each time from ObjectHandleOnStack. It would be more involved change.

Reply:
I think this shows that we should do the change in dotnet/runtime first so that subtle GC holes (if there are any) will show up as intermittent crashes in CI. The servicing CI runs much less often and so these bugs would take a lot longer to surface.

Reply:
I will do this on master first, I just can't turn this into draft, so I continued working here first.